src/crypto/slow-hash.c

```diff
@@ -1767,3 +1841,129 @@ void cn_slow_hash(const void *data, size_t length, char *hash, int variant, int
+    rx_sp->rs_height = height;
+    if (miners) {
+      if (rx_dataset == NULL) {
+        rx_dataset = randomx_alloc_dataset(RANDOMX_FLAG_LARGE_PAGES);
+        if (rx_dataset == NULL)
+          rx_dataset = randomx_alloc_dataset(RANDOMX_FLAG_DEFAULT);
+      }
+      if (rx_dataset != NULL) {
+        if (miners > 1) {
+          unsigned long delta = randomx_dataset_item_count() / miners;
+          unsigned long start;
+          int i;
+          seedinfo *si;
+          THREAD_TYPE *st;
+          si = malloc(miners * sizeof(seedinfo));
+          if (si == NULL)
+            goto leave;
+          st = malloc(miners * sizeof(THREAD_TYPE));
+          if (st == NULL) {
+            free(si);
+            goto leave;
+          }
+          for (i=0; i<miners; i++) {
+              si[i].si_start = start;
+            si[i].si_count = delta;
+            start += delta;
+          }
+          si[i-1].si_count = randomx_dataset_item_count() - si[i-1].si_start;
+          for (i=1; i<miners; i++) {
+            THREAD_CREATE(st[i], rx_seedthread, &si[i]);
+          }
+          randomx_init_dataset(rx_dataset, rx_sp->rs_cache, 0, si[0].si_count);
+          for (i=1; i<miners; i++) {
+            THREAD_JOIN(st[i]);
+          }
+          free(st);
+          free(si);
+        } else {
+          randomx_init_dataset(rx_dataset, rx_sp->rs_cache, 0, randomx_dataset_item_count());
+        }
+        flags |= RANDOMX_FLAG_FULL_MEM;
+      }
+    }
+    if (rx_vm == NULL) {
+      if(!force_software_aes() && check_aes_hw())
+        flags |= RANDOMX_FLAG_HARD_AES;
+      rx_vm = randomx_create_vm(flags | RANDOMX_FLAG_LARGE_PAGES, rx_sp->rs_cache, rx_dataset);
+      if(rx_vm == NULL) //large pages failed
+        rx_vm = randomx_create_vm(flags, rx_sp->rs_cache, rx_dataset);
+      if(rx_vm == NULL) {//fallback if everything fails
+        flags = RANDOMX_FLAG_DEFAULT | (miners ? RANDOMX_FLAG_FULL_MEM : 0);
+        rx_vm = randomx_create_vm(flags, rx_sp->rs_cache, rx_dataset);
+      }
+    } else {
+      randomx_vm_set_cache(rx_vm, rx_sp->rs_cache);
+    }
+  }
+leave:
+  MUTEX_UNLOCK(rx_mutex);
+}
+
+void rx_slow_hash(const void *data, size_t length, char *hash) {
+  if (rx_vm == NULL) {
+    rx_state *rx_sp = &rx_s[rx_s_toggle];
+    randomx_flags flags = RANDOMX_FLAG_DEFAULT;
+    if (use_v4_jit())
+      flags |= RANDOMX_FLAG_JIT;
+    if(!force_software_aes() && check_aes_hw())
+      flags |= RANDOMX_FLAG_HARD_AES;
+    rx_vm = randomx_create_vm(flags | RANDOMX_FLAG_LARGE_PAGES, rx_sp->rs_cache, NULL);
+    if(rx_vm == NULL) //large pages failed
+      rx_vm = randomx_create_vm(flags, rx_sp->rs_cache, NULL);
+    if(rx_vm == NULL) {//fallback if everything fails
+      flags = RANDOMX_FLAG_DEFAULT;
+      rx_vm = randomx_create_vm(flags, rx_sp->rs_cache, NULL);
+    }
+  }
+  if (rx_vm != NULL) {
+    randomx_calculate_hash(rx_vm, data, length, hash);
+  } else {
+    memset(hash, 0, 32);
+  }
```

Comment:
That looks like it should be a hard local_abort, no ?

Reply:
Yes...